pipeline/transformer/transformer_mets/src/main/scala/uk/ac/wellcome/platform/transformer/mets/parsers/MetsXmlParser.scala

```diff
@@ -48,4 +50,96 @@ object MetsXmlParser {
           new Exception("Found multiple accessCondtions in METS XML"))
     }
   }
+
+  /** Here we use the first defined item in the physicalStructMap to lookup a
+    *  file ID the fileObjects mapping, and use the files location as the
+    *  thumbnail image.
+    */
+  private def thumbnailLocation(root: Elem): Option[String] =
+    physicalStructMap(root).headOption
+      .flatMap {
+        case (_, fileId) =>
+          fileObjects(root).get(fileId)
+      }
+      .map(_.stripPrefix("objects/"))
+
+  /** The METS XML contains locations of associated files, contained in a
+    *  mapping with the following format:
+    *
+    *  <mets:fileSec>
+    *    <mets:fileGrp USE="OBJECTS">
+    *      <mets:file ID="FILE_0001_OBJECTS" MIMETYPE="image/jp2">
+    *        <mets:FLocat LOCTYPE="URL" xlink:href="objects/b30246039_0001.jp2" />
+    *      </mets:file>
+    *      <mets:file ID="FILE_0002_OBJECTS" MIMETYPE="image/jp2">
+    *        <mets:FLocat LOCTYPE="URL" xlink:href="objects/b30246039_0002.jp2" />
+    *      </mets:file>
+    *    </mets:fileGrp>
+    *  </mets:fileSec>
+    *
+    *  We need to look up the physicalStructMap to ascertain what each of these
+    *  files refer to.
+    */
+  private def fileObjects(root: Elem): Map[String, String] =
+    (root \ "fileSec" \ "fileGrp")
+      .filterByAttribute("USE", "OBJECTS")
+      .childrenWithTag("file")
+      .toMapping(
+        keyAttrib = "ID",
+        valueNode = "FLocat",
+        valueAttrib = "{http://www.w3.org/1999/xlink}href"
+      )
+
+  /** Valid METS documents should a physicalStructMap section, with the bottom
+    *  most divs each representing a physical page, and linking to files in the
+    *  corresponding fileSec structures:
+    *
+    *  <mets:structMap TYPE="PHYSICAL">
+    *    <mets:div DMDID="DMDPHYS_0000" ID="PHYS_0000" TYPE="physSequence">
+    *      <mets:div ADMID="AMD_0001" ID="PHYS_0001" ORDER="1" TYPE="page">
+    *        <mets:fptr FILEID="FILE_0001_OBJECTS" />
+    *        <mets:fptr FILEID="FILE_0001_ALTO" />
+    *      </mets:div>
+    *      <mets:div ADMID="AMD_0002" ID="PHYS_0002" ORDER="2" TYPE="page">
+    *        <mets:fptr FILEID="FILE_0002_OBJECTS" />
+    *        <mets:fptr FILEID="FILE_0002_ALTO" />
+    *      </mets:div>
+    *    </mets:div>
+    *  </mets:structMap>
+    */
+  private def physicalStructMap(root: Elem): ListMap[String, String] =
+    (root \ "structMap")
+      .filterByAttribute("TYPE", "PHYSICAL")
+      .descendentsWithTag("div")
+      .toMapping(
+        keyAttrib = "ID",
+        valueNode = "fptr",
+        valueAttrib = "FILEID"
+      )
```

Comment:
It’s not clear to me which of the `mets:ftpr` tags this will retrieve. Do you get:

```scala
Map("PHYS_0001" -> "FILE_0001_OBJECTS", "PHYS_0002" -> "FILE_0002_OBJECTS")
```

or

```scala
Map("PHYS_0001" -> "FILE_0001_ALTO", "PHYS_0002" -> "FILE_0002_ALTO")
```

And does it matter which? What if these attributes appear in a different order?

I had a look at the docs ListMap; it wasn't clear to me what it's getting beyond `Map[String, String]`. What am I missing?

Reply:
`ListMap` is an implementation of `Map` which preserves ordering: relevant as we want to retrieve the first defined item in the `thumbnailLocation` method above.

We get the first `fptr` node (line 135 in the `toMapping` method).

I will add some comments to make things more clear

Reply:
> And does it matter which? What if these attributes appear in a different order?

Good question. It appears like the first one is what we want, but I don't know that we can guarantee that is always correct. Unfortunately, there doesn't seem to be other information which we can use (the only attribute being the ID). I suppose we could choose the one which exists in the `OBJECTS` map, which seems a bit crazy to me as indicates recipricol relationships between each of these structures, but maybe that is how it is.

[This METS specification](https://www.loc.gov/standards/mets/profiles/00000014.html) says the following:

> Each div element representing a page of the item(s) to be displayed should have one or more child fptr elements. One fptr element should appear for each of the different-sized images in the fileSec, referencing the ID of for the page described in the parent div in the FILEID attribute. Other fptr children may be present; however, METS Navigator will not use these in its display or functionality.